### PR TITLE
Rewrite function declarations

### PR DESCRIPTION
**Pull request by pamelafox**
### High-level description of change

This rewrites function declarations into variable declarations. It does the rewriting when all the other rewriting is done, which happens after linting and tests are run.

The goal is that I can write examples for AP CSP that use function declaration syntax, as that is more compatible with how Code.org students are learning functions. That makes it easier for Code.org students/teachers to use the KA materials for remote learning.

### Are there performance implications for this change?

It's one more check while walking the AST. Should be negligible?

### Have you added tests to cover this new/updated code?

Yes, I added several in ast-transform. Happy to add any more edge cases you can think of.

### Risks involved

I think the AP CSP examples would be fine/not risky, I can put this on a ZND to see.
However, If other people start using this syntax, they may run into unforeseen issues with it: interaction with real-time update, interaction with challenge graders, etc. If they do run into issues, I disabled the babyhint so they would not see any message that'd indicate why something went awry.

Do you think this could put programs at risk somehow that aren't using function declarations? (Are there nodes with type=FunctionDeclaration besides what I'm testing?)

### Are there any dependencies or blockers for merging this?

webapp would need to be updated upstream

### How can we verify that this change works?

Try this program, see rect and ellipse:

```
function dry(x, y) {
    function innerFunc(x, y) {
        ellipse(x, y, 16, 13);
    }
    rect(x, y, 167, 137);
    innerFunc(102, 88);
}

dry(120, 95);

```

### CODE REVIEW

File: tests/output/webpage/output_test.js [hunk 1]
```diff
@@ -263,24 +263,11 @@ describe("Linting", function() {
...
-
-    if (isFirefox()) {
-        // slowparse succeeds when parsing this HTML on Firefox.
-        failingTest("Not so fatal slowparse error detected (Firefox)",
-            "<li><a href='</li><img src='https://www.kasandbox.org'>", [
-                {row: 0, column: 7, lint: {type: "INVALID_ATTR_NAME"}}
-            ]
-        );
-    }
+    failingTest("Slowparse error detected",
```
Comment (pamelafox) [inline tests/output/webpage/output_test.js]: It used to be that Chrome and Firefox behaved differently, but they now appear to act the same (at least locally, let's see what TravisCI says)

**Comment by danielhollas** [comment]
FWIW that "slow parse error" test was also failing for me locally when I was working on #724. Glad to know I am not the only one. :slightly_smiling_face:

File: js/output/pjs/pjs-ast-transforms.js [hunk 2]
```diff
@@ -318,3 +318,29 @@ ASTTransforms.rewriteNewExpressions = function(envName) {
...
+                    name: node.id.name
+                },
+                init: {
+                    type: "FunctionExpression",
+                    id: null,
+                    params: node.params,
+                    body: node.body,
+                    generator: node.generator,
+                    expression: node.expression,
+                    async: node.async
```
Comment (kevinbarabash) [inline js/output/pjs/pjs-ast-transforms.js]: just in case 😉

File: js/output/pjs/babyhint.js [hunk 3]
```diff
@@ -218,8 +218,6 @@ var BabyHint = {
 
             // Checks could detect new errors, thus must run on every line
             errors = errors
-                // check for incorrect function declarations
-                .concat(BabyHint.checkFunctionDecl(line, lineNumber))
```
Comment (kevinbarabash) [inline js/output/pjs/babyhint.js]: It's nice that we don't have to disable all of BabyHint.

**Comment by pamelafox** [comment]
@danielhollas good to know! Must be some environment mismatch.

@kevinbarabash Thanks for the review! I deployed to a ZND and things are looking good there so far:  https://static-znd-200317-pamela-p.khanacademy.org/computer-programming/new/pjs

**Comment by danielhollas** [comment]
@pamelafox not sure if this is of any concerns for AP CSP student, but potentially confusing thing is that function hoisting does not work in this approach:

image

**Comment by pamelafox** [comment]
@danielhollas I'm fine with that, as I think it's better practice to define your functions first. (More cross-language friendly). But thank you for bringing it up, it's useful to track the effects of this change/approach.
